test/IdRegistry/IdRegistry.st.sol

```diff
@@ -0,0 +1,192 @@
+            idRegistry.disableTrustedOnly();
+        }
+        if (svm.createBool("pauseRegistration?")) {
+            idRegistry.pauseRegistration();
+        }
+    }
+
+    // Verify the IdRegistry invariants
+    function check_invariant(bytes4 selector, address caller) public {
+        init();
+
+        // Consider two distinct addresses
+        vm.assume(x != y);
+
+        // Record pre-state
+        uint256 oldIdX = idRegistry.idOf(x);
+        uint256 oldIdY = idRegistry.idOf(y);
+
+        uint256 oldIdCounter = idRegistry.getIdCounter();
+
+        bool oldPaused = idRegistry.paused();
+
+        // Execute an arbitrary tx to IdRegistry
+        vm.prank(caller);
+        (bool success,) = address(idRegistry).call(mk_calldata(selector));
+        vm.assume(success); // ignore reverting cases
+
+        // Record post-state
+        uint256 newIdX = idRegistry.idOf(x);
+        uint256 newIdY = idRegistry.idOf(y);
+
+        uint256 newIdCounter = idRegistry.getIdCounter();
+
+        // Verify invariant properties
+
+        // Ensure that there is no recovery address associated with fid 0.
+        assert(idRegistry.getRecoveryOf(0) == address(0));
+
+        // Ensure that idCounter never decreases.
+        assert(newIdCounter >= oldIdCounter);
+
+        // If a new fid is registered, ensure that:
+        // - IdRegistry must not be paused.
+        // - idCounter must increase by 1.
+        // - The new fid must not be registered for an existing fid owner.
+        // - The existing fids must be preserved.
+        if (newIdCounter > oldIdCounter) {
+            assert(oldPaused == false);
+            assert(newIdCounter - oldIdCounter == 1);
+            assert(newIdX == oldIdX || oldIdX == 0);
+            assert(newIdX == oldIdX || newIdY == oldIdY);
+        }
+    }
+
+    function mk_calldata(bytes4 selector) internal returns (bytes memory) {
+        // Generate calldata based on the function selector
+        bytes memory args;
+        if (selector == idRegistry.registerFor.selector) {
+            args = abi.encode(
+                svm.createAddress("to"),
+                svm.createAddress("recovery"),
+                svm.createUint256("deadline"),
+                svm.createBytes(65, "sig")
+            );
+        } else if (selector == idRegistry.transfer.selector) {
+            args = abi.encode(svm.createAddress("to"), svm.createUint256("deadline"), svm.createBytes(65, "sig"));
+        } else if (selector == idRegistry.recover.selector) {
+            args = abi.encode(
+                svm.createAddress("from"),
+                svm.createAddress("to"),
+                svm.createUint256("deadline"),
+                svm.createBytes(65, "sig")
+            );
+        } else {
+            args = svm.createBytes(1024, "data");
+        }
+        return abi.encodePacked(selector, args);
+    }
+
+    function check_transfer(address caller, address to, address other) public {
```

Comment:
we test transfer in two ways - via this specific test for transfer, and more generically in check_invariant, though we do not assert much there. 

we test register only in one way - in the general check invariant, where we assert a lot of specific stuff. 

is there a rationale for this design? or is this something that could be standardizes where we have a general invariant assertion test and specific ones for each function including register?